src/YDialogSpy.cc

```diff
@@ -278,70 +468,422 @@ void fillTree( YWidgetTreeItem * 		parent,
-		yuiDebug() << "Highlighting " << item << std::endl;
-
-		if ( item )
-		{
-		    priv->targetDialog->highlight( item->widget() );
-		    showProperties( item->widget() );
-		}
-	    }
-	}
+    return (item) ? item->widget() : 0;
+}
+
+/**
+ * The selected item has been changed, refresh the UI
+ */
+void YDialogSpyPrivate::selectedWidgetChanged()
+{
+    highlightWidget();
+    refreshProperties();
+    refreshButtonStates();
+}
+
+/**
+ * Run the property editor for the current widget
+ */
+void YDialogSpyPrivate::editProperty()
+{
+    auto selected_item = dynamic_cast<YTableItem *>(propTable->selectedItem());
+    if (!selected_item) return;
+
+    auto cell = selected_item->cell(0);
+    yuiMilestone() << "editing property: " << cell->label();
+
+    YPropertyEditor editor(selectedWidget());
+    // update the property table when only the property has been changed
+    if (editor.edit(cell->label())) refreshProperties();
+}
+
+/**
+ * Delete the currently selected widget
+ */
+void YDialogSpyPrivate::deleteWidget()
+{
+    auto w = selectedWidget();
+    if (!w) return;
+
+    auto parent = w->parent();
+    if (!parent) return;
+
+    yuiMilestone() << "removing widget: " << w << std::endl;
+    parent->removeChild(w);
+
+    if ( w->isValid() )
+    {
+        delete w;
     }
+
+    // any other child left after the removal?
+    if (!parent->hasChildren())
+    {
+        // add an Empty widget to have a valid widget tree
+        // e.g. empty VBoxes are not allowed
+        YUI::widgetFactory()->createEmpty(parent);
+    }
+
+    targetDialogUpdated();
 }
 
+/**
+ * Helper method - Is the widget a VBox or Hbox?
+ * @param  widget the widget
+ * @return        true if the widget is a VBox or HBox
+ */
+bool isBox(YWidget *widget)
+{
+    return dynamic_cast<YLayoutBox *>(widget);
+}
 
-void YDialogSpy::showDialogSpy( YDialog * dialog )
+/**
+ * Helper method - Is the widget a VBox?
+ * @param  widget the widget
+ * @return        true if the widget is a VBox
+ */
+bool isVBox(YWidget *widget)
+{
+    auto box = dynamic_cast<YLayoutBox *>(widget);
+    return box && box->primary() == YD_VERT;
+}
+
+/**
+ * Move the selected widget up/left or down/right. The visual direction
+ * actually depends on the widget, it just moves the widget to the begining
+ * or the end of the container.
+ * @param true = up move to the begining (up/left), false = to the end (down/right)
+ */
+void YDialogSpyPrivate::moveSelected(bool up)
+{
+    auto target_widget = selectedWidget();
+    if (!target_widget) return;
+
+    auto parent = target_widget->parent();
+    if (!parent || !isBox(parent)) return;
+
+    if (up)
+    {
+        // the first child cannot be moved further
+        if (target_widget == parent->firstChild()) return;
+
+        auto i = find( parent->childrenBegin(), parent->childrenEnd(), target_widget );
+        if (i != parent->childrenEnd())
+        {
+            // swap with the preceeding widget
+            std::swap(*(--i), *i);
```

Comment:
AFAIS this is undefined, see http://stackoverflow.com/questions/9566187/function-parameter-evaluation-order.